Sales forecasting end to end

goyetc opened the pull request:
Moved folder to root of truera-examples repo. Updated readme based on Josh's feedback.

rohitrastogi commented:
Wow, this is a great resource - it is a very comprehensive view of how to use our APIs.

Say you weren't using diagnostics to begin with - what would be your plan of attack for ingesting a dataset purely so you have a baseline to compare production data for drift calculations? Would you expect a customer to already have predictions stored somewhere so perhaps they wouldn't need the model object at all? Or, would you expect them to use the model object to generate the predictions?

Also, how strictly do customers follow the flow you took for Diagnostics?
1. upload features/labels 
2. train a model 
3. use model artifact to generate predictions and fis for each model
4. ingest predictions and fis for the models 
5. Repeat 2-4

Are we always used directly within the model training/experimentation process, or do customers also use us after models are already fit?

rohitrastogi commented on end-to-end_sales-forecasting/python_scripts/sales_forecasting_1_create_project_dev_preds_FI.py:
```diff
@@ -0,0 +1,208 @@
+    
+tru.set_model("Ridge Regression")
+tru.set_data_split("training data")
+lr_train_feat_infs = tru.compute_feature_influences(stop=len(X_train_pre))
```
Are the influences generated on the pre data or post data?

rohitrastogi replied:
Does diagnostics show the post data features in the features page?

goyetc replied:
I probably should have used len(X_train_post) -- that is the data that is model executable / is used, under the covers, for the .compute execution. In practice it makes no difference, as they're both the same length. But will change/update.

There is a feature map implemented in this workflow, so the features page / all results will be based on pre-transform. In this case, this basically just amounts to showing categorical vars with multiple levels, versus the total set of binary responses of M categorics x N(M) levels.

rohitrastogi commented:
Also, at what step between 1-5, would you expect a customer to look at the Diagnostics UI?

joshreini1 commented on end-to-end_sales-forecasting/sales_forecasting-end-to-end.ipynb:
```diff
@@ -0,0 +1,2701 @@
+{
```
<p>Can we download from pypi rather than whl on downloads page?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/truera-examples/pull/21/discussion/#comment-1358540602'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

goyetc replied:
<p>Yes, will update this. </p>

joshreini1 commented on end-to-end_sales-forecasting/sales_forecasting-end-to-end.ipynb:
```diff
@@ -0,0 +1,2701 @@
+{
```
<p>nit: please condense inputs into one or two cells</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/truera-examples/pull/21/discussion/#comment-1358540617'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='1'/>

goyetc replied:
<p>sure, can do</p>

joshreini1 commented on end-to-end_sales-forecasting/sales_forecasting-end-to-end.ipynb:
```diff
@@ -0,0 +1,2701 @@
+{
```
<p>Do we need so much eda here? Might be better to get to using TruEra sooner</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/truera-examples/pull/21/discussion/#comment-1358540623'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

goyetc replied:
<p>I removed most of the EDA in a prior iteration. The steps that remain are requisite for processing the raw data to a state that it can be modeled and ingested by TruEra. In the interest of this actually being an 'end to end' notebook, I'd like to leave the remaining steps in. </p>

joshreini1 commented on end-to-end_sales-forecasting/sales_forecasting-end-to-end.ipynb:
```diff
@@ -0,0 +1,2701 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #43.            ## 2022 update: this probably is no longer necessary, either because timestamps are included as index in data files, and/or that we will no longer require manual set of begin &amp; end timestamps in near future, for monitoring ingestion
</pre><p>Can we just remove the timestamp as extra data if its no longer necessary (and this comment)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/truera-examples/pull/21/discussion/#comment-1358540635'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='39'/>

goyetc replied:
<p>Done</p>

joshreini1 commented on end-to-end_sales-forecasting/sales_forecasting-end-to-end.ipynb:
```diff
@@ -0,0 +1,2701 @@
+{
```
<p>Seems like some formatting off with this cell?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/truera-examples/pull/21/discussion/#comment-1358540645'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='178'/>

goyetc replied:
<p>Purposefully in raw format -- if you want to sample the data and produce FIs for &lt; len(prod data), then users can use that cell </p>

joshreini1 commented on end-to-end_sales-forecasting/sales_forecasting-end-to-end.ipynb:
```diff
@@ -0,0 +1,2701 @@
+{
```
<p>Let's remove the appendix from the polished notebook</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/truera-examples/pull/21/discussion/#comment-1358540657'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='185'/>

goyetc replied:
<p>10-4</p>

joshreini1 commented:
Looks like you currently have the end2end folder in both the quickstart and root folder.

Can you also add this to the README in the root?

goyetc commented:
> Also, at what step between 1-5, would you expect a customer to look at the Diagnostics UI?

likely after step 1 -- at that point, the development data and models would be fully populated. 

In an alternative scenario, one might choose to use elements of script 5 to ingest existing data (i.e., all I/Os available, including pre-computed FIs) for whichever data is available to them.

goyetc commented:
> Looks like you currently have the end2end folder in both the quickstart and root folder.
Removed 
> Can you also add this to the README in the root?
Will do

avinashmaheshjoshi commented:
> Wow, this is a great resource - it is a very comprehensive view of how to use our APIs.
> 
> Say you weren't using diagnostics to begin with - what would be your plan of attack for ingesting a dataset purely so you have a baseline to compare production data for drift calculations? Would you expect a customer to already have predictions stored somewhere so perhaps they wouldn't need the model object at all? Or, would you expect them to use the model object to generate the predictions?
> 
> Also, how strictly do customers follow the flow you took for Diagnostics?
> 
> 1. upload features/labels
> 2. train a model
> 3. use model artifact to generate predictions and fis for each model
> 4. ingest predictions and fis for the models
> 5. Repeat 2-4
> 
> Are we always used directly within the model training/experimentation process, or do customers also use us after models are already fit?


@goyetc Would be interested to know about your replies on @rohitrastogi's above questions

My view is that,

We are used after the model is fit. 

1. For simple models, we have the option for the user to ingest the pre-data and then use our compute functions to generate predictions and influences and then upload them (without having to define the schema etc...)

2. For complex models or cases where the user doesn't want to upload the model binary for whatever reasons (IP protection, too large etc...), we allow users to ingest predictions and influences computed by them using external functions and then use add_data or the specific functions like add_model_predictions or model_model_feature influences to upload

The idea being by being flexible, this should cover all the use-cases.

goyetc commented:
@Avinash Joshi ***@***.***>,

Rohit and I synced on Friday. Your view is in line with our discussion.

Some additional questions came up, including

   - Do users typically add a model or add data, first? Does this
   matter/are there reasons to do one before the other?
   - In the latter case you describe, (pre-existing I/Os or complex model
   not conducive to TruEra execution), how can we make this easier for users?
      - 'virtual model' concept -- how can this be more user-friendly /
      less jargon heavy?
      - what additional utilities can we provide in SDK to help users, e.g.,
      - (1) utility function to create or recall column spec
      - (2) utility function to creat or recall model output context
      - (3) more help or clarity on how or why to create a feature map
   - A discussion of the differences between a pre-prod and prod
   ingestion scenario, wrt what inputs will (almost always) be available
      1. pre-prod/Diagnostics: previously, we hardened the software to
      accommodate scenarios where labels are not available. In a
sense, this was
      because we were promoting the use of Diagnostics for the purposes of
      production monitoring. We also did not require predictions, in the prior,
      model-runner-based world.
      2. Now that we have Monitoring -- is there ever a scenario where a
      split loaded *directly* into Diagnostics would not have labels? In
      other words, in current state of software, is the only way a label-less
      split would enter Diagnostics via the Monitoring-->Diagnostics workflow,
      via a time range split? It seems so.
      3. There appear to be opportunities for streamlining the SDK /
      ingestion process based on (2) above.

cc @Rohit Rastogi ***@***.***> for further discussion.

Colin

--
TruEra | +1 (845) 242-7070



On Mon, Oct 16, 2023 at 5:54 AM Avinash Joshi ***@***.***>
wrote:

> Wow, this is a great resource - it is a very comprehensive view of how to
> use our APIs.
>
> Say you weren't using diagnostics to begin with - what would be your plan
> of attack for ingesting a dataset purely so you have a baseline to compare
> production data for drift calculations? Would you expect a customer to
> already have predictions stored somewhere so perhaps they wouldn't need the
> model object at all? Or, would you expect them to use the model object to
> generate the predictions?
>
> Also, how strictly do customers follow the flow you took for Diagnostics?
>
>    1. upload features/labels
>    2. train a model
>    3. use model artifact to generate predictions and fis for each model
>    4. ingest predictions and fis for the models
>    5. Repeat 2-4
>
> Are we always used directly within the model training/experimentation
> process, or do customers also use us after models are already fit?
>
> @goyetc <https://github.com/goyetc> Would be interested to know about
> your replies on @rohitrastogi <https://github.com/rohitrastogi>'s above
> questions
>
> My view is that,
>
> We are used after the model is fit.
>
>    1.
>
>    For simple models, we have the option for the user to ingest the
>    pre-data and then use our compute functions to generate predictions and
>    influences and then upload them (without having to define the schema etc...)
>    2.
>
>    For complex models or cases where the user doesn't want to upload the
>    model binary for whatever reasons (IP protection, too large etc...), we
>    allow users to ingest predictions and influences computed by them using
>    external functions and then use add_data or the specific functions like
>    add_model_predictions or model_model_feature influences to upload
>
> The idea being by being flexible, this should cover all the use-cases.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/truera/truera-examples/pull/21#issuecomment-1764121502>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFLYBEDHCY44YTPJS3VWAZTX7T74PANCNFSM6AAAAAA5SZOENE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
